### PR TITLE
Dala 399 internal storage

### PR DESCRIPTION
**Pull request by stewal19**
# Pull Request \<Title>
`<Description here>`
## Things to do during Peer Review
Please check all boxes before the Pull Request is merged. In case you skip a box, describe in the PRs description (that means: here) why the check is skipped.
- [x] The Github Actions (including Sonarqube Gateway and Lint Checks) are green. This is enforced by Github. 
- [x] A peer-review has been executed
  - [x] The code has been manually inspected by someone who did not implement the feature
- [x] The PR actually implements what is described in the JIRA-Issue
- [x] At least one E2E Test exists testing the new feature
- [x] Documentation is updated as required
- [x] The automated deployment is updated if required
- [x] The new version is deployed to the dev server "dev1" using this branch
  - [x] It's verified that this version actually is the one deployed (check actuator/info for branch name and commit id!)
  - [x] The new feature is manually used/tested/observed on dev server
  - [x] All implemented Social Logins have been tested manually in the UI
- [x] If any work on the UI is to be merged, those changes were also documented in the Figma 
- [x] The local Dev stack still works: execute `startDevelopmentStack.sh` and execute `npm run testpipeline -- --env EXECUTION_ENVIRONMENT=""` in dataland-frontend
- [x] After(!) the cypress tests have passed locally, execute the backend-e2e-tests `./gradlew dataland-e2etests:test`

### CODE REVIEW

File: .github/workflows/CD.yaml [hunk 1]
```diff
@@ -194,8 +190,8 @@ jobs:
         run: cd dataland-frontend && npm install && npm run testgithubenvironment --url="https://${{ secrets.TARGETSERVER_URL }}"
         env:
           ELECTRON_EXTRA_LAUNCH_ARGS: "--ignore-connections-limit=dataland.com,test.dataland.com,dev1.dataland.com,dev2.dataland.com"
-          CYPRESS_PREPOPULATE_TIMEOUT_S: ${{ github.event.inputs.useInMemoryStorage && 180 || 900 }}
-          CYPRESS_PREVISIT_TIMEOUT_S: ${{ github.event.inputs.useInMemoryStorage && 180 || 1200 }}
+          CYPRESS_PREPOPULATE_TIMEOUT_S: ${{ 900 }}
+          CYPRESS_PREVISIT_TIMEOUT_S: ${{ 1200 }}
```
Comment (ppuetsch) [inline .github/workflows/CD.yaml:194]: 900/1200 are needed for EuroDaT - as this took ages. 
For any internal service, the shorter values (180/180) should do the job.

File: dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/services/DatabaseDataStore.kt [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return dataID
+    }
+
+    /**
+     * Reads data from a database
+     * @param dataId the id of the data to be retrieved
+     * @return the data as json string with id dataId
+     */
+    fun selectDataSet(dataId: String): String {
+        return dataItemRepository.findById(dataId).orElse(DataItem("", "", "")).data
```
Comment (ppuetsch) [inline dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/services/DatabaseDataStore.kt:35]: Here we could log something - including the correlationId

File: dataland-internal-storage/src/main/resources/application.properties [hunk 3]
```diff
@@ -0,0 +1,22 @@
+springdoc.show-actuator=true
+management.endpoints.web.exposure.include=health, info
+management.endpoint.health.show-details=always
+management.info.git.mode=FULL
+server.tomcat.threads.min-spare=100
+server.shutdown=graceful
+server.servlet.context-path=/internal-storage
+spring.security.oauth2.resourceserver.jwt.issuer-uri=https://${PROXY_PRIMARY_URL:local-dev.dataland.com}/keycloak/realms/datalandsecurity
+spring.security.oauth2.resourceserver.jwt.jwk-set-uri=http://keycloak:8080/keycloak/realms/datalandsecurity/protocol/openid-connect/certs
```
Comment (ppuetsch) [inline dataland-internal-storage/src/main/resources/application.properties:9]: spring security not used - hence the properties can be deleted.

File: dataland-keycloak-adapter/build.gradle.kts [hunk 4]
```diff
@@ -19,6 +19,9 @@ plugins {
     jacoco
     id("org.springframework.boot")
     kotlin("kapt")
+
+    id("com.gorylenko.gradle-git-properties")
+    id("org.jetbrains.kotlin.plugin.jpa")
```
Comment (ppuetsch) [inline dataland-keycloak-adapter/build.gradle.kts:24]: why was this added?

File: dataland-internal-storage/src/main/resources/application-development.properties [hunk 5]
```diff
@@ -0,0 +1,5 @@
+spring.security.oauth2.resourceserver.jwt.issuer-uri=https://local-dev.dataland.com/keycloak/realms/datalandsecurity
+spring.security.oauth2.resourceserver.jwt.jwk-set-uri=https://local-dev.dataland.com/keycloak/realms/datalandsecurity/protocol/openid-connect/certs
```
Comment (ppuetsch) [inline dataland-internal-storage/src/main/resources/application-development.properties:2]: spring security not used - hence the properties can be deleted.

File: dataland-pgadmin/pgpass.template [hunk 6]
```diff
@@ -1,3 +1,4 @@
 keycloak-db:5432:*:keycloak:${KEYCLOAK_DB_PASSWORD}
 backend-db:5432:*:backend:${BACKEND_DB_PASSWORD}
-api-key-manager-db:5432:*:api_key_manager:${API_KEY_MANAGER_DB_PASSWORD}
\ No newline at end of file
+api-key-manager-db:5432:*:api_key_manager:${API_KEY_MANAGER_DB_PASSWORD}
+internal-storage-db:5432:*:internal_storage:${INTERNAL_STORAGE_DB_PASSWORD}
```
Comment (ppuetsch) [inline dataland-pgadmin/pgpass.template:4]: newline at End of File is missing

File: deployment/start_and_deploy_to_server.sh [hunk 7]
```diff
@@ -4,17 +4,7 @@ set -euxo pipefail
...
-else
-  profile=production
-  echo "Checking if EuroDaT is available before deploying to target server."
-  if ! curl -f -X 'GET' "${TRUSTEE_BASE_URL}/${TRUSTEE_ENVIRONMENT_NAME}/api/check/health" -H 'accept: application/json' >/dev/null 2>&1; then
-    echo "EuroDaT is not available."
-    exit 1
-  fi
-  echo "EuroDat is available."
-fi
+profile=productionInMemory
```
Comment (ppuetsch) [inline deployment/start_and_deploy_to_server.sh:7]: why should profile be `productionInMemory`?

File: docker-compose.yml [hunk 8]
```diff
@@ -180,6 +165,38 @@ services:
       EDC_OAUTH_TOKEN_URL: ${EDC_OAUTH_TOKEN_URL}
       IDS_WEBHOOK_ADDRESS: ${IDS_WEBHOOK_ADDRESS}
       SERVER_TOMCAT_THREADS_MIN-SPARE: 100
+  internal-storage:
+    restart: always
+    profiles: [ "production", "productionInMemory", "testing", "development" ]
```
Comment (ppuetsch) [inline docker-compose.yml:170]: can't `productionInMemory` go away?

File: dataland-backend/build.gradle.kts [hunk 9]
```diff
@@ -30,6 +30,8 @@ dependencies {
     implementation(project(":dataland-backend-utils"))
     implementation(libs.springdoc.openapi.ui)
     implementation(libs.dataland.edc.client)
+    implementation(libs.moshi.kotlin)
+    implementation(libs.moshi.adapters)
```
Comment (ppuetsch) [inline dataland-backend/build.gradle.kts:34]: why are those added?

File: dataland-internal-storage/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+    kotlin("kapt")
+    id("org.jetbrains.kotlin.plugin.jpa")
+    id("org.openapi.generator")
+}
+
+java.sourceCompatibility = JavaVersion.VERSION_17
+
+dependencies {
+    implementation(libs.springdoc.openapi.ui)
+    implementation(libs.okhttp)
```
Comment (ppuetsch) [inline dataland-internal-storage/build.gradle.kts:34]: is okhttp really needed?

File: dataland-internal-storage/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    implementation(libs.logback.classic)
+    implementation(libs.logback.core)
+    implementation(libs.slf4j.api)
+    implementation("org.springframework.boot:spring-boot-starter-web")
+    implementation("org.springframework.boot:spring-boot-starter-actuator")
+    implementation("org.springframework.boot:spring-boot-starter-data-jpa")
+    runtimeOnly(libs.database.postgres)
+    runtimeOnly(libs.database.h2)
+    implementation(libs.bcpkix.jdk15on)
+    implementation(libs.bcprov.jdk15on)
```
Comment (ppuetsch) [inline dataland-internal-storage/build.gradle.kts:47]: are those  (`libs.bcp*`) needed?

File: testing/wait_for_backend_and_run_e2etests.sh [hunk 12]
```diff
@@ -7,7 +7,7 @@ set -euxo pipefail
 is_infrastructure_up () {
```
Comment (ppuetsch) [inline testing/wait_for_backend_and_run_e2etests.sh:7]: do we actually need this anymore? Isn't it obsolete by waiting for docker images to be healthy?

File: testing/wait_for_backend_and_run_e2etests.sh [hunk 13]
```diff
@@ -23,7 +23,7 @@ is_infrastructure_up () {
 }
 export -f is_infrastructure_up
 
-timeout 240 bash -c "while ! is_infrastructure_up; do echo 'infrastructure not yet completely there - retrying in 1s'; sleep 1; done; echo 'infrastructure up!'"
+timeout 340 bash -c "while ! is_infrastructure_up; do echo 'infrastructure not yet completely there - retrying in 1s'; sleep 1; done; echo 'infrastructure up!'"
```
Comment (ppuetsch) [inline testing/wait_for_backend_and_run_e2etests.sh:26]: for what do we need the extra 100s?

File: build.gradle.kts [hunk 14]
```diff
@@ -5,6 +5,8 @@ val ktlintVersion: String by project
 val githubUser: String by project
 val githubToken: String by project
 
+extra["internalStorageOpenApi"] = "internalStorageOpenApi.json"
```
Comment (ppuetsch) [inline build.gradle.kts:8]: why do we need this _here_? It's not needed for the other files?

File: build.gradle.kts [hunk 15]
```diff
@@ -65,6 +67,7 @@ plugins {
     id("org.openapi.generator") version "6.2.1" apply false
     id("com.github.ben-manes.versions") version "0.43.0"
     id("org.jetbrains.kotlin.plugin.jpa") version "1.7.21" apply false
+    id("io.swagger.core.v3.swagger-gradle-plugin") version "2.2.6" apply false
```
Comment (ppuetsch) [inline build.gradle.kts:70]: why do we need that? We didn't need it for the other openapi files

File: dataland-keycloak-adapter/build.gradle.kts (hunk 4 above)
Comment (fsoffel) [reply dataland-keycloak-adapter/build.gradle.kts:24]: I don't know :I will be removed

File: docker-compose.yml (hunk 8 above)
Comment (fsoffel) [reply docker-compose.yml:170]: Will be checked.

File: testing/wait_for_backend_and_run_e2etests.sh (hunk 12 above)
Comment (fsoffel) [reply testing/wait_for_backend_and_run_e2etests.sh:7]: The e2etests container needs a way to check if the other services are up before. There is currently no other function implemented that provides that check from an external view and this file is executed at the entrypoint of the e2etests dockerfile where for some reason no healthcheck exists. However, we could remove the execution of the script from the entrypoint add an always true health check and start the script from outside the container. What do you think?

File: testing/wait_for_backend_and_run_e2etests.sh (hunk 13 above)
Comment (fsoffel) [reply testing/wait_for_backend_and_run_e2etests.sh:26]: debug fragment

File: build.gradle.kts (hunk 14 above)
Comment (fsoffel) [reply build.gradle.kts:8]: We use it for client generation in the backend.

File: .github/workflows/CD.yaml (hunk 1 above)
Comment (ppuetsch) [reply .github/workflows/CD.yaml]: As 180 is anyways set in `dataland-frontend/cypress.env.json`, we should remove this variable everywhere. What do you thing?

File: dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/controller/StorageController.kt [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+class StorageController(
+    @Autowired val dataStore: DatabaseDataStore
+) : StorageAPI {
+
+    override fun selectDataById(dataId: String, correlationId: String?): ResponseEntity<String> {
+        return ResponseEntity.ok(dataStore.selectDataSet(dataId))
+    }
+
+    override fun insertData(correlationId: String?, body: String?): ResponseEntity<InsertDataResponse> {
+        // TODO Question for the review process: is "" an ok default body?
```
[the remark this answers is not in the document]
Comment (ppuetsch) [reply dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/controller/StorageController.kt]: ok

File: dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/services/DatabaseDataStore.kt [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+    /**
+     * Insterts data into a database
+     * @param correlationId id of the action inserting this data
+     * @param data a json object
+     * @return id associated with the stored data
+     */
+    fun insertDataSet(correlationId: String?, data: String): String {
+        val dataID = "${UUID.randomUUID()}:${UUID.randomUUID()}_${UUID.randomUUID()}"
+        dataItemRepository.save(DataItem(dataID, correlationId ?: "", data))
```
[the remark this answers is not in the document]
Comment (ppuetsch) [reply dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/services/DatabaseDataStore.kt]: ok

File: dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/services/DatabaseDataStore.kt (hunk 2 above)
Comment (ppuetsch) [reply dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/services/DatabaseDataStore.kt]: ok

File: dataland-internal-storage/src/main/resources/application.properties (hunk 3 above)
Comment (ppuetsch) [reply dataland-internal-storage/src/main/resources/application.properties]: ok

File: dataland-keycloak-adapter/build.gradle.kts (hunk 4 above)
Comment (ppuetsch) [reply dataland-keycloak-adapter/build.gradle.kts]: ok

File: dataland-internal-storage/src/main/resources/application-development.properties (hunk 5 above)
Comment (ppuetsch) [reply dataland-internal-storage/src/main/resources/application-development.properties]: ok

File: dataland-pgadmin/pgpass.template (hunk 6 above)
Comment (ppuetsch) [reply dataland-pgadmin/pgpass.template]: ok

File: deployment/start_and_deploy_to_server.sh (hunk 7 above)
Comment (ppuetsch) [reply deployment/start_and_deploy_to_server.sh]: it's production now.

File: docker-compose.yml (hunk 8 above)
Comment (ppuetsch) [reply docker-compose.yml]: it's away now.

File: dataland-internal-storage/build.gradle.kts (hunk 11 above)
Comment (ppuetsch) [reply dataland-internal-storage/build.gradle.kts]: gone now.

File: testing/wait_for_backend_and_run_e2etests.sh (hunk 12 above)
Comment (ppuetsch) [reply testing/wait_for_backend_and_run_e2etests.sh:7]: let's leave as ist.

Comment (ppuetsch) [reply testing/wait_for_backend_and_run_e2etests.sh:7]: at least for now ;-)

File: build.gradle.kts (hunk 14 above)
Comment (ppuetsch) [reply build.gradle.kts]: ok

File: build.gradle.kts (hunk 15 above)
Comment (ppuetsch) [reply build.gradle.kts]: ok

File: testing/wait_for_backend_and_run_e2etests.sh (hunk 13 above)
Comment (ppuetsch) [reply testing/wait_for_backend_and_run_e2etests.sh]: ok

File: dataland-internal-storage/build.gradle.kts (hunk 10 above)
Comment (ppuetsch) [reply dataland-internal-storage/build.gradle.kts]: it's gone now

File: dataland-internal-storage/src/main/resources/application-nodb.properties [hunk 18]
```diff
@@ -0,0 +1,7 @@
+spring.security.oauth2.resourceserver.jwt.issuer-uri=https://local-dev.dataland.com/keycloak/realms/datalandsecurity
+spring.security.oauth2.resourceserver.jwt.jwk-set-uri=https://local-dev.dataland.com/keycloak/realms/datalandsecurity/protocol/openid-connect/certs
```
[the remark this answers is not in the document]
Comment (ppuetsch) [reply dataland-internal-storage/src/main/resources/application-nodb.properties]: ok

File: dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/entities/DataItem.kt [hunk 19]
```diff
@@ -0,0 +1,21 @@
...
+ * The database entity for storing company data
+ */
+@Entity
+@Table(name = "data_items")
+data class DataItem(
+    @Id
+    @Column(name = "data_id")
+    val id: String,
+    @Column(name = "correlation_id")
+    val correlationId: String,
```
[the remark this answers is not in the document]
Comment (ppuetsch) [reply dataland-internal-storage/src/main/kotlin/org/dataland/datalandinternalstorage/entities/DataItem.kt]: ok

File: .github/workflows/CD.yaml (hunk 1 above)
Comment (fsoffel) [reply .github/workflows/CD.yaml]: For the current state where we do not use EuroDaT, we would not need to set them here. The variables CYPRESS_PREVISIT_TIMEOUT_S and CYPRESS_PREPOPULATE_TIMEOUT_S seem not to be used right now, but PREVISIT_TIMEOUT_S and PREPOPULATE_TIMEOUT_S via Cypress.env. I assume the prefix is automatically added before the lookup then? Are the constants from these file automatically applied? Then there would be more redundant constants defined, like AWAIT_PREPOPULATION_RETRIES, I'll look that up and update the code.

Comment (fsoffel) [reply .github/workflows/CD.yaml]: It seems reasonable to remove one ore the other pair. I'll remove the CYPRESS_* variables for now. Although they will be probably more handy when dealing with EuroDaT again. What irritates me: Have they even been of use recently? Or have they been overwritten by the ones from cypress.env.json all the time.

Comment (fsoffel) [reply .github/workflows/CD.yaml]: The other variables defined in that file make it appear that, the variables in cypress.env.json are only applied (if at all) on the developer machine. In that case we would need to set them in the docker-compose, right?
